### PR TITLE
API catch up

### PR DESCRIPTION
**Pull request by mustberuss**
API Changes since PR #27 

-   The 10,000 row/10 page limit seems to have gone away.  Should we warn if the return will be over 1M rows or so? 

-   Five of the [endpoints](https://search.patentsview.org/swagger-ui/) are now nested under patent/  The easiest thing to do was to have users qualify the nested endpoints but there might be a better/cleaner way? Ex: endpoint = "patent/us_application_citations".  

-  Removed nber_categories and nber_subcategories endpoints.  They were removed from the OpenAPI object but the API still responds to requests.

-  location endpoint is back

-   The cpc endpoints and their returned fields have new names.

 -  The new wipo endpoint returns a wipo entity, breaking the singular endpoint/plural entity pattern followed everywhere else.

  - The API currently throws 500 errors when all fields are requested from some of the endpoints.  I added a skip in test-search-pv.R.  

-   Paging is completely different.  Now the attribute is "after", where we supply the last value of the previous response.  This makes having a sort field required, I've added a default sort and deprecated the page parameter.  I haven't tried paging with a secondary sort, I'm not sure the API team has thought this through! (what value should "after" have when there is more than a primary sort) See Pagination for the ElasticSearch API (v0.1) here https://patentsview.org/apis/api-query-language#options_parameter

### CODE REVIEW

File: R/search-pv.R [hunk 1]
```diff
@@ -96,31 +95,35 @@ request_apply <- function(ex_res, method, query, base_url, arg_list, api_key, ..
   if (req_pages < 1) {
     stop2("No records matched your query...Can't download multiple pages")
   }
-  if (matched_records > 10000) {
```
Comment (crew102) [inline R/search-pv.R]: Are we sure that this limitation is going to be dropped permanently? I think I saw something on the docs that said that there *would* be a limitation, which makes me think that the lack of limitation is just a temporary thing.

File: tests/testthat/test-search-pv.R [hunk 2]
```diff
@@ -128,25 +132,28 @@ test_that("We can call all the legitimate HATEOAS endpoints", {
...
 
   single_item_queries <- c(
-    "https://search.patentsview.org/api/v1/assignee/10/",
-    "https://search.patentsview.org/api/v1/cpc_group/A01B/",
-    "https://search.patentsview.org/api/v1/cpc_subgroup/G01S7:4811/",
-    "https://search.patentsview.org/api/v1/cpc_subsection/A01/",
-    "https://search.patentsview.org/api/v1/inventor/10/",
-    "https://search.patentsview.org/api/v1/nber_category/1/",
-    "https://search.patentsview.org/api/v1/nber_subcategory/11/",
+    "https://search.patentsview.org/api/v1/assignee/00000ce5-b13f-4a23-a8fb-c14409ad7b68/",
```
Comment (crew102) [inline tests/testthat/test-search-pv.R]: I'm wondering if this repetition in the URL can be made more DRY? Maybe a small function that pastes the first part of the URL with the variable second?

File: R/search-pv.R [hunk 3]
```diff
@@ -96,31 +95,35 @@ request_apply <- function(ex_res, method, query, base_url, arg_list, api_key, ..
...
-      "how to get around this limitation."
-    )
-  }
 
   tmp <- lapply(seq_len(req_pages), function(i) {
-    arg_list$opts$offset <- (i - 1) * arg_list$opts$size
     x <- one_request(method, query, base_url, arg_list, api_key, ...)
     x <- process_resp(x)
+
+    # now to page we need set the "after" attribute to where we left off
```
Comment (crew102) [inline R/search-pv.R]: This seems like such a weird pattern, where the paging requires us to get the last ID that is sorted on in a given iteration. Any idea if this is a permanent thing? I'd be curious to know if the ElasticSearch DB is supporting this kind of paging natively or if it's hacked together on patentsviews' end. Any idea? I couldn't find the source code for the latest API code easily.

File: data-raw/openAPI_to_csv.pl [hunk 4]
```diff
@@ -0,0 +1,193 @@
+#!/usr/local/bin/perl
```
Comment (crew102) [inline data-raw/openAPI_to_csv.pl]: Can you summarize what's changed here from the R script?

**Comment by crew102** [comment]
Hey @mustberuss , thanks or the PR. One high-level comment - it looks like some of the examples need to be updated (e.g., some still using patent_number). Can you update them? I turned the running of examples back on during the Github CI build, so they should be run automatically now (and presumably would fail the build if they don't all run successfully).

File: R/search-pv.R (hunk 3 above)
Comment (mustberuss) [reply R/search-pv.R]: I agree it's a weird pattern!  I also looked for the source but couldn't find it. 

They did update the query language page mentioning the after parameter and the api throws errors if "offset" is passed so I'm guessing it's permanent.   ElasticSearch is new to me  but it looks like after is [supported natively](https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html#search-after).   The example shows passing multiple sort values when there is more than a primary sort which is something not currently handled in search-pv.   

ElasticSearch supports paging by offset so I don't see why they would make such a drastic change.  Perhaps we can down vote this API change!

File: tests/testthat/test-search-pv.R (hunk 2 above)
Comment (mustberuss) [reply tests/testthat/test-search-pv.R]: Sure, that's a good idea.  I'm wondering if it should be changed to make an ordinary query first and then use the links the API returns- so the test wouldn't break if the inventor, assignee and locations ids aren't persistent.

File: R/search-pv.R (hunk 1 above)
Comment (mustberuss) [reply R/search-pv.R]: I'm not sure if it's permanent or temporary.  They've been slow to fix reported bugs and haven't announced changes in advance.  I usually notice changes when test cases start failing or they update the OpenAPI spec (I make head requests to see if it's change recently).

**Comment by mustberuss** [comment]
@crew102  didn't you have your API key set here?  [My push](https://github.com/mustberuss/patentsview/actions/runs/3803203309) worked but failed here saying there wasn't an API key.

File: data-raw/openAPI_to_csv.pl (hunk 4 above)
Comment (mustberuss) [reply data-raw/openAPI_to_csv.pl]: The R script parses my copy of the OpenAPI definition from the previous iteration.  The perl script parses their updated definition directly https://search.patentsview.org/static/openapi.json.  Just changing the url in the R script throws the error below and I wasn't able to fix it.

Error in rownames_to_column(.) : is.data.frame(df) is not TRUE
Calls: extract_relevant_schema_info ... mutate -> setNames -> rownames_to_column -> stopifnot
Execution halted

File: R/search-pv.R (hunk 3 above)
Comment (mustberuss) [reply R/search-pv.R]: Paging does seem to work with a secondary sort, passing an array of both keys in the "after" parameter.

curl -X 'POST' \
  'https://search.patentsview.org/api/v1/patent/' \
  -H 'accept: application/json' \
  -H 'X-Api-Key: my-api-key' \
  -H 'Content-Type: application/json' \
  -d '{
  "f": ["patent_id", "patent_date"],
  "o": {"size": 100, "after":["2007-01-02", "RE39455"]},
  "q": {"_gte":{"patent_date":"2007-01-02"}},
  "s": [{"patent_date": "asc"}, {"patent_id": "asc"}]
}'

**Comment by crew102** [comment]
> @crew102 didn't you have your API key set here? [My push](https://github.com/mustberuss/patentsview/actions/runs/3803203309) worked but failed here saying there wasn't an API key.

Yeah apparently forks of a repo can't read secrets from the repo it's forked from.

**Comment by crew102** [comment]
Hey Russ, can you make sure the various checks in `check_query(query, endpoint)` work (i.e., let's uncomment that and fix any issues that arise).

Also I noticed that the endpoints are singular in the online docs (e.g., patent endpoint, not patents). Should we not coincide with the official names?

File: tests/testthat/test-search-pv.R [hunk 5]
```diff
@@ -90,27 +98,27 @@ test_that("Throttled requests are automatically retried", {
   skip_on_cran()
 
   res <- search_pv('{"_gte":{"patent_date":"2007-01-04"}}', per_page = 50)
-  patent_numbers <- res$data$patents$patent_number
+  patent_ids <- res$data$patents$patent_id
```
Comment (crew102) [inline tests/testthat/test-search-pv.R]: Is this test actually testing what we want it to now? Seems like there's no possible throttling happening now.

**Comment by mustberuss** [comment]
> Also I noticed that the endpoints are singular in the online docs (e.g., patent endpoint, not patents). Should we not coincide with the official names?

The returned entities are still plural (except wipo)  I tried making the endpoints singular but it  made the code more complex.  It was simpler to have the endpoints and returned entities match up.  I was also trying to maintain backward compatibility.  It's another of the API's changes I don't necessarily agree with!  It was like this in the mvp but obviously we can change it.

File: R/check-query.R [hunk 6]
```diff
@@ -42,11 +43,13 @@ check_query <- function(query, endpoint) {
...
 
-  flds_flt <- fieldsdf[fieldsdf$endpoint == endpoint & fieldsdf$can_query == "y", ]
+  flds_flt <- fieldsdf[fieldsdf$endpoint == endpoint, ]
 
   apply_checks <- function(x, endpoint) {
     x <- swap_null_nms(x)
-    if (names(x) %in% c("_not", "_and", "_or") || is.na(names(x))) {
+
+    # troublesome next line:  'length(x) = 2 > 1' in coercion to 'logical(1)'
+    if (length(names(x)) > 1 || names(x) %in% c("_not", "_and", "_or") || is.na(names(x))) {
```
Comment (mustberuss) [inline R/check-query.R]: The original line threw errors when checking queries.   Wasn't sure if this was the proper fix.

File: tests/testthat/test-search-pv.R (hunk 5 above)
Comment (mustberuss) [reply tests/testthat/test-search-pv.R]: The API is throttling, I updated the test to prove it.

**Comment by mustberuss** [comment]
> > @crew102 didn't you have your API key set here? [My push](https://github.com/mustberuss/patentsview/actions/runs/3803203309) worked but failed here saying there wasn't an API key.
> 
> Yeah apparently forks of a repo can't read secrets from the repo it's forked from.

Would it work if you set a secret in this repo?  I thought we had tests working here, or was that before they added API keys?

File: R/search-pv.R (hunk 3 above)
Comment (mustberuss) [reply R/search-pv.R]: That same ElasticSearch page says 10,000 rows is the paging limit when using the native size and from attributes.  It looks like the after parameter gets around that which must be why the API team made the change.  Guessing the API change will be permanent.

File: tests/testthat/test-search-pv.R (hunk 2 above)
Comment (mustberuss) [reply tests/testthat/test-search-pv.R]: I pushed [6d09a0b](https://github.com/ropensci/patentsview/pull/29/commits/6d09a0b58aa11c307beaf67918ac791c88b5adf8) to DRY out the code and query for links instead of hard coding ids that may not persist.

**Comment by crew102** [comment]
> > Yeah apparently forks of a repo can't read secrets from the repo it's forked from.
> 
> Would it work if you set a secret in this repo? I thought we had tests working here, or was that before they added API keys?

K I added you as a contributor. Can you try again? I'll take a look at the PR sometime this weekend.

**Comment by mustberuss** [comment]
> K I added you as a contributor. Can you try again? I'll take a look at the PR sometime this weekend.

Wow, thanks!  Can you merge in the change to _pkgdown.yml mentioned in #30?  I'm guessing Maëlle would see another failure in r-universe if I retry without it.

**Comment by crew102** [comment]
> > K I added you as a contributor. Can you try again? I'll take a look at the PR sometime this weekend.
> 
> Wow, thanks! Can you merge in the change to _pkgdown.yml mentioned in #30? I'm guessing Maëlle would see another failure in r-universe if I retry without it.

I can but let's create a separate PR for that. Can you open real quick and I'll merge.
